### ISSUE TITLE
Random Forest: Vastly different results to scikit learn

### ISSUE THREAD

**Issue by sjwhitworth** [sjwhitworth/golearn#84]
When training a random forest, I see vastly different, and poorer results when using golearn, than when using Python's scikit-learn. Unfortunately the dataset is confidential, so I can't share it here online. However, I'm using the same train/test split, and have ensured that data is represented the same way (they're all floats in both). 

When using scikit learn
```
Auc: 0.943867958106
Confusion Matrix
[[35878  1876]
 [ 5402 16388]]
             precision    recall  f1-score   support

          0       0.87      0.95      0.91     37754
          1       0.90      0.75      0.82     21790

avg / total       0.88      0.88      0.88     59544
```

When using golearn, with the same number of estimators..

```
Reference Class	True Positives	False Positives	True Negatives	Precision	Recall	F1 Score
---------------	--------------	---------------	--------------	---------	------	--------
1.00		4199		1366		15858		0.7545		0.4263	0.5448
0.00		15858		5651		4199		0.7373		0.9207	0.8188
Overall accuracy: 0.7408
```

As you can see, there's a big drop in precision and recall, on both outcomes. Any ideas as to what could be the problem @Sentimentron ?

**Comment by Aakash282** [comment]
hey guys I'm looking into using golearn for stuff at work and was wondering if there have been improvements on decision tree/RF predictive accuracy.

**Comment by sjwhitworth** [comment]
I don't think there has been any work on it thus far @Aakash282. Feel free to give it a shot and potentially compare - I can't think of any reason why they would differ.

**Comment by Aakash282** [comment]
Sorry, I should clarify my question. This issue was regarding the difference in performance between golearn RF implementation and sklearns implementation. It seems like last year there was a significant difference in the performance and I'm wondering if thats still true. I'm just trying to figure out whats viable for my project at work. Thanks!

**Comment by thedadams** [comment]
I think I may have stumbled across a similar problem.  I was using data that were all floats and using KNN.  Similar to @sjwhitworth, my outcomes were either 1 or 0.  I was getting outrageously accurate predictions.  I decided to change the outcomes in my dataset to "Hit" and "No Hit" and now I am getting extremely different results.

I think that the algorithm is mistakenly using the outcome as part of the learning and comparing process if it is a float.  Perhaps that is what is happening above: the algorithm is using the 0 and 1 as data as well as outcomes.

**Comment by Sentimentron** [comment]
In your case Donnie, it may be that there's a flaw in KNN's
`canUseOptimisations` function which may mean it's inappropriately using a
vectorized implementation that reads the class values. Could you try
setting `AllowOptimisations` to false in the `KNN` structure and seeing if
it displays the same issue?

**Comment by thedadams** [comment]
I switched `AllowOptimisations` to false, but the problem persists.

The data I have makes me fairly certain that the 0 and 1 outcomes are used in the prediction because the difference between 0 and 1 is greater than the difference in the other data.  Meaning that if the last column of the .csv file is being used in prediction, then that column will dominate when using KNN.  I am going from an overall accuracy of  96.86% using 0 and 1 for outcomes to an overall accuracy of 57.23% using outcomes of "Hit" and "No Hit".

**Comment by Sentimentron** [comment]
I managed to replicate the problem, with the Iris dataset, but I found the mismatch was because I was unwittingly read in the column headers as ordinary data. Make sure your loading statement looks like 
```go
rawData, err := base.ParseCSVToInstances("../examples/datasets/iris_headers_twoclass.csv", true)
```
Rather than
```go
rawData, err := base.ParseCSVToInstances("../examples/datasets/iris_headers_twoclass.csv", false)
```

Also note that if you're using `InstancesTrainTestSplit`, this doesn't return the same data each time because of random number generation. I've got a new patch which adds a deterministic version coming shortly.

**Comment by thedadams** [comment]
My data files don't have headers so I have set the second argument to `false`.  Also, I am not using `InstancesTrainTestSplit` because my training data and test data are in different files.

Is there something else I can show you or do to track down the problem?

**Comment by thedadams** [comment]
Here are two examples (different from the ones I gave above).  These are the same data except the outcomes in one are 0 and 1:
```
Reference Class	True Positives	False Positives	True Negatives	Precision	Recall	F1 Score
---------------	--------------	---------------	--------------	---------	------	--------
1.000		113		0		51		1.0000		1.0000	1.0000
0.000		51		0		113		1.0000		1.0000	1.0000
Overall accuracy: 1.0000
```

The other, the outcomes are "Hit" and "No-Hit":
```
Reference Class	True Positives	False Positives	True Negatives	Precision	Recall	F1 Score
---------------	--------------	---------------	--------------	---------	------	--------
No-Hit		3		8		105		0.2727		0.0588	0.0968
Hit		105		48		3		0.6863		0.9292	0.7895
Overall accuracy: 0.6585
```

I have a fix for `knn.go` and I will submit a pull request, if that is ok with the collaborators.

**Comment by Sentimentron** [comment]
Ah, yes, I see the problem and it's really quite obvious is that the un-optimised version is not correctly filtering based on class attributes. I'll move discussion to that PR.

**Comment by njern** [comment]
Fixed by merging #118
